kong/tools/http.lua

```diff
@@ -0,0 +1,587 @@
+-- terminated). It will check in the current request (global `ngx` table). If
+-- the header `X-Forwarded-Proto` exists -- with value `https` then it will also
+-- be considered as an https connection.
+-- @param trusted_ip boolean indicating if the client is a trusted IP
+-- @param allow_terminated if truthy, the `X-Forwarded-Proto` header will be checked as well.
+-- @return boolean or nil+error in case the header exists multiple times
+_M.check_https = function(trusted_ip, allow_terminated)
+  if ngx.var.scheme:lower() == "https" then
+    return true
+  end
+
+  if not allow_terminated then
+    return false
+  end
+
+  -- if we trust this IP, examine it's X-Forwarded-Proto header
+  -- otherwise, we fall back to relying on the client scheme
+  -- (which was either validated earlier, or we fall through this block)
+  if trusted_ip then
+    local scheme = ngx.req.get_headers()["x-forwarded-proto"]
+
+    -- we could use the first entry (lower security), or check the contents of
+    -- each of them (slow). So for now defensive, and error
+    -- out on multiple entries for the x-forwarded-proto header.
+    if type(scheme) == "table" then
+      return nil, "Only one X-Forwarded-Proto header allowed"
+    end
+
+    return tostring(scheme):lower() == "https"
+  end
+
+  return false
+end
+
+
+local CONTROLS = [[\x00-\x1F\x7F]]
+local HIGHBIT = [[\x80-\xFF]]
+local SEPARATORS = [==[ \t()<>@,;:\\\"\/?={}\[\]]==]
+local HTTP_TOKEN_FORBID_PATTERN = "[".. CONTROLS .. HIGHBIT .. SEPARATORS .. "]"
+
+
+--- Validates a token defined by RFC 2616.
+-- @param token (string) the string to verify
+-- @return the valid token, or `nil+error`
+function _M.validate_http_token(token)
+  if token == nil or token == "" then
+    return nil, "no token provided"
+  end
+
+  if not re_match(token, HTTP_TOKEN_FORBID_PATTERN, "jo") then
+    return token
+  end
+
+  return nil, "contains one or more invalid characters. ASCII " ..
+              "control characters (0-31;127), space, tab and the " ..
+              "characters ()<>@,;:\\\"/?={}[] are not allowed."
+end
+
+
+-- should we also use validate_http_token for this?
+--- Validates a header name.
+-- Checks characters used in a header name to be valid, as per nginx only
+-- a-z, A-Z, 0-9 and '-' are allowed.
+-- @param name (string) the header name to verify
+-- @return the valid header name, or `nil+error`
+function _M.validate_header_name(name)
+  if name == nil or name == "" then
+    return nil, "no header name provided"
+  end
+
+  if re_match(name, "^[a-zA-Z0-9-_]+$", "jo") then
+    return name
+  end
+
+  return nil, "bad header name '" .. name ..
+              "', allowed characters are A-Z, a-z, 0-9, '_', and '-'"
+end
+
+
+--- Validates a cookie name.
+-- @param name (string) the cookie name to verify
+-- @return the valid cookie name, or `nil+error`
+_M.validate_cookie_name = _M.validate_http_token
+
+
+local validate_labels
```

Comment:
It seems that `validate_labels` is used by conf_loader, it does not belong to http.

Reply:
Yes, it seems I overlooked this. I will create a new ticket to address it. The code has been moved back to `kong/tools/utils`.